include/flang/lower/ASTBuilder.h

```diff
@@ -0,0 +1,332 @@
+  static void setBranches(Evaluation *cstr) {
+    if (cstr) {
+      cstr->setBranches();
+    }
+  }
+
+  EvalVariant u;
+  ParentType parent;
+  parser::CharBlock pos;
+  std::optional<parser::Label> lab;
+  std::list<Evaluation> *subs{nullptr}; // construct sub-statements
+  CFGAnnotation cfg{CFGAnnotation::None};
+  bool isTarget{false};         // this evaluation is a control target
+  bool containsBranches{false}; // construct contains branches
+};
+
+/// A program is a list of program units.
+/// These units can be function like, module like, or block data
+struct ProgramUnit {
+  template <typename A>
+  ProgramUnit(A *ptr, const ParentType &parent) : p{ptr}, parent{parent} {}
+
+  std::variant<const parser::MainProgram *, const parser::FunctionSubprogram *,
+               const parser::SubroutineSubprogram *, const parser::Module *,
+               const parser::Submodule *,
+               const parser::SeparateModuleSubprogram *,
+               const parser::BlockData *>
+      p;
+  ParentType parent;
+};
+
+/// Function-like units have similar structure. They all can contain executable
+/// statements.
+struct FunctionLikeUnit : public ProgramUnit {
+  // wrapper statements for function-like syntactic structures
+  using FunctionStatement =
+      std::variant<const parser::Statement<parser::ProgramStmt> *,
+                   const parser::Statement<parser::EndProgramStmt> *,
+                   const parser::Statement<parser::FunctionStmt> *,
+                   const parser::Statement<parser::EndFunctionStmt> *,
+                   const parser::Statement<parser::SubroutineStmt> *,
+                   const parser::Statement<parser::EndSubroutineStmt> *,
+                   const parser::Statement<parser::MpSubprogramStmt> *,
+                   const parser::Statement<parser::EndMpSubprogramStmt> *>;
+
+  FunctionLikeUnit(const parser::MainProgram &f, const ParentType &parent);
+  FunctionLikeUnit(const parser::FunctionSubprogram &f,
+                   const ParentType &parent);
+  FunctionLikeUnit(const parser::SubroutineSubprogram &f,
+                   const ParentType &parent);
+  FunctionLikeUnit(const parser::SeparateModuleSubprogram &f,
+                   const ParentType &parent);
+
+  bool isMainProgram() {
+    return std::get_if<const parser::Statement<parser::EndProgramStmt> *>(
+        &funStmts.back());
+  }
+  const parser::FunctionStmt *isFunction() {
+    return isA<parser::FunctionStmt>();
+  }
+  const parser::SubroutineStmt *isSubroutine() {
+    return isA<parser::SubroutineStmt>();
+  }
+  const parser::MpSubprogramStmt *isMPSubp() {
+    return isA<parser::MpSubprogramStmt>();
+  }
+
+  const semantics::Scope *scope{nullptr}; // scope from front-end
+  std::list<FunctionStatement> funStmts;  // begin/end pair
+  std::list<Evaluation> evals;            // statements
+  std::list<FunctionLikeUnit> funcs;      // internal procedures
+
+private:
+  template <typename A>
+  const A *isA() {
```

Comment:
asA might be a better name for this than isA (isA in LLVM just does a boolean check it doesn't do a cast)

Reply:
The intention is really that of a getter.

Reply:
The name was updated to `getA`.